File: node.js/lib/cerrio/cerrio.js
var Connection = require('./connection');
var Stream = require('./stream');

module.exports = Cerrio;

function Cerrio(port) {  
    this.connection = new Connection(port);

    this.subscribe = function(options) {
        var stream = new Stream(this.connection, options);
        this.connection.onActive(function() { stream.start(); }, true);
        return stream;
    }

    this.stop = function() {
    	this.connection.stop();
    }
    
    var connection = this.connection;

    this.add = function(uri, data) {
        if (!uri) {
            throw new Error("uri can't be null");
        }
        var sourceId = this.sourceId;
        connection.onActive(function() {
            connection.socket.emit('update', {
                id: sourceId + ":add",
                uri: uri,
                update: {
                    action: 'add',
                    item: data
                }
            });
        }, false);
    };

    this.modify = function(uri,data) {
    	if(!uri){
    		throw new Error("uri can't be null");
    	}
        var sourceId = this.sourceId;
        connection.onActive(function(){
            connection.socket.emit('update', {
                id: sourceId + ":modify",
                uri: uri,
                update: {
                    action: 'modify',
                    item: data
                }
            });
        },false);
    };

    this.sendDelete = function(uri,key) {
    	if(!uri){
    		throw new Error("uri can't be null");
    	}
        var sourceId = this.sourceId;
        connection.onActive(function(){
            connection.socket.emit('update', {
                id: sourceId + ":delete",
                uri: uri,
                update: {
                    action: 'delete',
                    itemKey: key
                }
            });
        },false);
    };
    
    this.applySession = function(token){
    	connection.applySession(token);
    };
}

File: java/src/ly/xstream/streaming/Cerrio.java
package ly.xstream.streaming;

class Cerrio {
	public Connection connection;
	private ILogger logger;
	
	public Cerrio(int port, ILogger logger){
		connection = new Connection(port,logger);
		this.logger = logger;
	}
	
	public Stream subscribe(SubscriptionOptions options){
		Stream stream = new Stream(connection,options,logger);
		StreamStarter streamStarter = new StreamStarter(stream);
		connection.onActive(streamStarter, true);
		return stream;
	}
	
	public void stop(){
		connection.stop();
	}
	
	void sendAdd(String uri,Object data){
		connection.onActive(new DataSender(new CerrioUpdateWrapper("add",uri,connection.id,data,null)), false);
	}
	
	void sendModify(String uri,Object data){	
		connection.onActive(new DataSender(new CerrioUpdateWrapper("modify",uri,connection.id,data,null)), false);
	}
	
	void sendModify(String uri,Object data,String... modifiedFields){	
		connection.onActive(new DataSender(new CerrioUpdateWrapper("modify",uri,connection.id,data,null,modifiedFields)), false);
	}
	
	void sendDelete(String uri,String itemKey){
		connection.onActive(new DataSender(new CerrioUpdateWrapper("delete",uri,connection.id,null,itemKey)), false);
	}
	
	public void applySession(String token){
		connection.applySession(token);
	}
	
	private class DataSender implements IConnectionActiveCallback{
		
		private CerrioUpdateWrapper wrapper;
		
		public DataSender(CerrioUpdateWrapper wrapper){
			this.wrapper = wrapper;
		}
		
		@Override
		public void onConnectionActive(Connection connection) {
			try {
				connection.send("update", wrapper);
			} catch (Exception e) {
				logger.handleError("problem sending data: ",e);
			}
		}
		
	}
	
	private class StreamStarter implements IConnectionActiveCallback{
		Stream stream;
		
		public StreamStarter(Stream stream){
			this.stream = stream;
		}
		
		
		@Override
		public void onConnectionActive(Connection connection) {
			stream.start();
		}
	}
}
